Add support for loading dashboards in orgs

hervenicol opened the pull request:
### What this PR does / why we need it

Towards https://github.com/giantswarm/roadmap/issues/3696

Support loading dashboards in organizations.

Extract from README telling about this feature:

> It will look for kubernetes `ConfigMaps` and use them as dashboards if they meet these criteria:
> - a label `app.giantswarm.io/kind: "dashboard"`
> - an annotation or label `giantswarm.io/organization` set to the organization the dasboard should be loaded in.
> 
> Current limitations:
> - no support for folders
> - each dashboard belongs to one and only one organization


### Checklist

- [x] Update changelog in CHANGELOG.md.
- [ ] Follow deployment test procedure in the tests/manual_e2e directory and have a working branch.

TheoBrigitte commented on internal/controller/dashboard_controller.go:
```diff
@@ -0,0 +1,344 @@
+/*
+Copyright 2024.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
```
There's no need for this, we already have the license at the repository root

hervenicol replied:
This is is all controllers, so I'd rather remove them in a separate PR.

hervenicol replied:
=> https://github.com/giantswarm/observability-operator/pull/213

TheoBrigitte replied:
Still it should be removed here also

TheoBrigitte commented on internal/controller/dashboard_controller.go:
```diff
@@ -0,0 +1,344 @@
+}
+
+const (
+	DashboardFinalizer      = "observability.giantswarm.io/grafanadashboard"
+	DashboardTeamAnnotation = "giantswarm.io/organization"
```
The organization label should be re-use from there https://github.com/giantswarm/observability-operator/blob/d61b8b68b92520d5633b66bdb0502c4c449be991/pkg/common/organization/repository.go#L13
feel free to export it by making it capital case

TheoBrigitte commented on internal/controller/dashboard_controller.go:
```diff
@@ -0,0 +1,344 @@
+		return errors.WithStack(err)
+	}
+
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&v1.ConfigMap{}, builder.WithPredicates(labelSelectorPredicate)).
```
Adding a name to the controller otherwise it default to the reconciled resources, so here it would be name configmap in the logs.
```suggestion
	return ctrl.NewControllerManagedBy(mgr).
		Named("dashboards")
		For(&v1.ConfigMap{}, builder.WithPredicates(labelSelectorPredicate)).
```

TheoBrigitte commented on internal/controller/dashboard_controller.go:
```diff
@@ -0,0 +1,344 @@
+// SetupWithManager sets up the controller with the Manager.
+func (r *DashboardReconciler) SetupWithManager(mgr ctrl.Manager) error {
+
+	labelSelectorPredicate, err := predicate.LabelSelectorPredicate(metav1.LabelSelector{MatchLabels: map[string]string{"app.giantswarm.io/kind": "dashboard"}})
```
Please move the label name and value somewhere in a constant.

TheoBrigitte commented on internal/controller/dashboard_controller.go:
```diff
@@ -0,0 +1,344 @@
+				for _, dashboard := range dashboards.Items {
+					requests = append(requests, reconcile.Request{
+						NamespacedName: types.NamespacedName{
+							Name: dashboard.Name,
```
I assume dashboard have a namespace since they are a configmap, the namespace should be added here.

TheoBrigitte commented on internal/controller/dashboard_controller.go:
```diff
@@ -0,0 +1,344 @@
+		// Watch for grafana pod's status changes
+		Watches(
+			&v1.Pod{},
+			handler.EnqueueRequestsFromMapFunc(func(ctx context.Context, obj client.Object) []reconcile.Request {
```
I would prefer to have this to be defined in a separate func block rather than embedded here.

TheoBrigitte replied:
Yeah ok for me

TheoBrigitte commented on internal/controller/dashboard_controller.go:
```diff
@@ -0,0 +1,344 @@
+
+// DashboardReconciler reconciles a Dashboard object
+type DashboardReconciler struct {
+	client.Client
```
I find it confusing to have client being embedded here, I would rather have it as a regular field but I'am not strong on this one.

hervenicol replied:
That's the same for all reconcilers, I don't think changing this should be part of the current PR.

TheoBrigitte replied:
Ok

TheoBrigitte commented on internal/controller/dashboard_controller.go:
```diff
@@ -0,0 +1,344 @@
+
+func getDashboardUID(jsonDashboard string) (UID string, err error) {
+
+	var Dashboard map[string]interface{}
+	err = json.Unmarshal([]byte(jsonDashboard), &Dashboard)
+	if err != nil {
+		return "", err
+	}
+
+	UID, ok := Dashboard["uid"].(string)
```
That's an easy workaround, but if you only need to get the id you could also declare a minimal dashboard type like this


```go
type dashboard struct {
	ID `json:"id"`
}
```
and use it here as
```go
var Dashboard dashboard
```

But that's just a suggestion, I'am not strong on this one

TheoBrigitte commented on internal/controller/dashboard_controller.go:
```diff
@@ -0,0 +1,344 @@
+func getDashboardCMOrg(dashboard *v1.ConfigMap) (Org string, err error) {
+
+	// Try to look for an annotation first
+	orgName := dashboard.GetAnnotations()[DashboardTeamAnnotation]
+	if orgName != "" {
+		return orgName, nil
+	}
+
+	// Then look for a label
+	orgName = dashboard.GetLabels()[DashboardTeamAnnotation]
```
Please check for return values from GetAnnotations and GetLabels, as it migh be nil

TheoBrigitte replied:
I meant that `GetAnnotations` and `GetLabels` both return a map, and map can be `nil` so before accessing the value we should check if the map is not nil otherwise this would cause a panic.

Something like this
```go
	labels := dashboard.GetLabels()
	if labels != nil && labels[organization.OrganizationLabel] != "" {
		return labels[organization.OrganizationLabel], nil
	}
```

hervenicol replied:
oh, I thought the "ok" check was preventing this to happen.
All right, I'll do it as you recommend! Looks more explicit!

TheoBrigitte commented on internal/controller/dashboard_controller.go:
```diff
@@ -0,0 +1,344 @@
+	return ctrl.Result{}, nil
+}
+
+func getDashboardUID(jsonDashboard string) (UID string, err error) {
```
I don't think there's a need for named return arguments here

hervenicol replied:
Is it bad practice to used named return arguments? Should I try to avoid it?

TheoBrigitte commented on internal/controller/dashboard_controller.go:
```diff
@@ -0,0 +1,344 @@
+	return UID, nil
+}
+
+func getDashboardCMOrg(dashboard *v1.ConfigMap) (Org string, err error) {
```
I don't think there's a need for named return arguments here

TheoBrigitte commented on internal/controller/dashboard_controller.go:
```diff
@@ -0,0 +1,344 @@
+			logger.Error(err, "failed to find organization", "organization", dashboardOrg)
+			return errors.WithStack(err)
+		}
+		logger.Info("Found dashboard org", "Organization", organization)
```
This log line would be verbose and a similar information is already logged later on

TheoBrigitte commented on internal/controller/dashboard_controller.go:
```diff
@@ -0,0 +1,344 @@
+			FolderUID: "",
+			IsFolder:  false,
+			Message:   "Added by observability-operator",
+			Overwrite: true,
```
What's the dashboard identifier used to overwrite it ? Is it by name ? Please document this.

TheoBrigitte commented on internal/controller/dashboard_controller.go:
```diff
@@ -0,0 +1,344 @@
+	}
+
+	// return nil
+	return errors.New("All good, but not implemented yet")
```
Please don't merge this.

hervenicol replied:
:scream: :face_with_open_eyes_and_hand_over_mouth:

TheoBrigitte reviewed (commented):
Nice job, sorry for all the remarks :see_no_evil:

hervenicol commented:
> Nice job, sorry for all the remarks 🙈

Thanks!
I did a first round of fixes, and I added a " :+1: " to each remark I fixed.
Feel free to check them and for each remark either "Resolve" or add new comments if you consider my fixes were not good enough @TheoBrigitte  :bow: 

Then I'll keep going on with the remaining remarks.

TheoBrigitte commented:
> > Nice job, sorry for all the remarks 🙈
> 
> Thanks! I did a first round of fixes, and I added a " 👍 " to each remark I fixed. Feel free to check them and for each remark either "Resolve" or add new comments if you consider my fixes were not good enough @TheoBrigitte 🙇
> 
> Then I'll keep going on with the remaining remarks.

Did a round of review on the reviews, closed a couple and commented where needed

hervenicol commented:
Added dashboards deletion and a bit of rework:
- change org only once per configmap (instead as for each dashboard in the CM)
- in case of error for a dashboard, continue reconciling next dashboards for this configmap

QuentinBisson commented on README.md:
```diff
@@ -7,6 +7,18 @@ This operator is in charge of handling the setup and configuration of the Giant
 
+## Features
+
+### Grafan dashboards provisioning
```
```suggestion
### Grafana dashboards provisioning
```

QuentinBisson commented on README.md:
```diff
@@ -7,6 +7,18 @@ This operator is in charge of handling the setup and configuration of the Giant
+
+It will look for kubernetes `ConfigMaps` and use them as dashboards if they meet these criteria:
+- a label `app.giantswarm.io/kind: "dashboard"`
+- an annotation or label `giantswarm.io/organization` set to the organization the dasboard should be loaded in.
```
I think we use this label for giant swarm organizations so we probably should not use this but smth like observability.giantswarm.io/organization instead. WDYT?

hervenicol replied:
WDYM "giantswarm organizations"?
I used the existing const `organizationLabel` in `pkg/common/organization/repository.go` but it may very well be wrong.

hervenicol replied:
Right, I just understood that the constant I was using was not related to grafana config at all.
Then I wondered where to put this constant, but for now the dashboards controller is the only one using it, so let's just keep the constant in it.

Done.

QuentinBisson commented on internal/controller/dashboard_controller.go:
```diff
@@ -0,0 +1,354 @@
+		Cert: conf.Environment.GrafanaTLSCertFile,
+		Key:  conf.Environment.GrafanaTLSKeyFile,
+	}
+	grafanaAPI, err := grafanaclient.GenerateGrafanaClient(conf.GrafanaURL, grafanaAdminCredentials, grafanaTLSConfig)
```
We have the same code here https://github.com/giantswarm/observability-operator/blob/69c4b70780fa0d4e202a3f8b6c1712506d203a1d/internal/controller/grafanaorganization_controller.go#L44. Maybe we should create a function or create the client in main?

hervenicol replied:
Each controller must have its own grafana client, because this client uses a context with current org for instance.
And when one controller changes this context, we don't want it to have an impact on what another controller does.

But we could probably have a function that reads the creds, creates a grafanaclient and returns it. yes :+1:

hervenicol replied:
Where should it be? Probably not in `pkg/grafana/client` if we want it to remain generic, right?
Should we have something like `internal/controller/common` then?

QuentinBisson replied:
I think something like `internal/controller/clients/grafana.go` might be better semantically, WDYT?

TheoBrigitte replied:
I think we have no need to use the special `internal` package path, I would rather have this in the already existing `pkg/grafana/client` module, avoiding to split things into too many different places.

QuentinBisson replied:
Can you move the whole credentials and tls part in the function as well?

hervenicol replied:
done.

QuentinBisson commented on internal/controller/dashboard_controller.go:
```diff
@@ -0,0 +1,354 @@
+}
+
+func getDashboardUID(dashboard map[string]interface{}) (string, error) {
+
```
Nitpick but I don't think we need those blank lines :)

hervenicol replied:
Reminds me of a discussion with @TheoBrigitte where we were wondering about coding style guidelines :grin: 

So, you mean a new block should not start with a blank line?
I've got lots of functions and `for` blocks that start with a blank line. If I was to do the cleanup, I'd do it all around then.

TheoBrigitte replied:
I confirm, blocks usually do not start with blank lines.

hervenicol replied:
All right, should be fixed for my current PR's changes.

QuentinBisson commented on internal/controller/dashboard_controller.go:
```diff
@@ -0,0 +1,354 @@
+		}
+
+		// Create or update dashboard
+		_, err = r.GrafanaAPI.Dashboards.PostDashboard(&models.SaveDashboardCommand{
```
Could we move that to the grafana pkg and not the controller?

hervenicol replied:
Just did it, please check it's as you expect.

QuentinBisson commented on internal/controller/dashboard_controller.go:
```diff
@@ -0,0 +1,345 @@
+	}()
+
+	// Switch context to the dashboards-defined org
+	organization, err := grafana.FindOrgByName(r.GrafanaAPI, dashboardOrg)
```
Here, I'm not sure if the name should be the GrafanaOrganization CR `metadata.name` as this field cannot change (at least not without recreating the whole org) or GrafanaOrganization CR `spec.displayName` as this value can be updated by users.

hervenicol replied:
I asked myself the same question.

The displayName (as used currently) seems more flexible, as it's not tied to the way we create orgs.
It allows to load dashboards in orgs that don't have a CR, ie orgs that were created via other ways like clickops.
In the future, we can change the way we create the orgs without breaking dashboards loading.

QuentinBisson replied:
sure

QuentinBisson commented on internal/controller/dashboard_controller.go:
```diff
@@ -0,0 +1,345 @@
+
+		dashboardUID, err := getDashboardUID(dashboard)
+		if err != nil {
+			logger.Info("Skipping dashboard, no UID found")
```
Same question about logger.Error

QuentinBisson commented on internal/controller/dashboard_controller.go:
```diff
@@ -0,0 +1,345 @@
+		}
+
+		// Create or update dashboard
+		err = grafana.SetDashboard(r.GrafanaAPI, dashboard)
```
NIT :D 
```suggestion
		err = grafana.PublishDashboard(r.GrafanaAPI, dashboard)
```

QuentinBisson commented on internal/controller/dashboard_controller.go:
```diff
@@ -0,0 +1,345 @@
+
+	dashboardOrg, err := getDashboardCMOrg(dashboardCM)
+	if err != nil {
+		logger.Info("Skipping dashboard, no organization found")
```
logger.Error would be best

QuentinBisson commented on internal/controller/dashboard_controller.go:
```diff
@@ -0,0 +1,345 @@
+			continue
+		}
+
+		// TODO: search for dashboard by ID
```
Do we need the todos?

hervenicol replied:
Nope, cleaning them up.

QuentinBisson commented on main.go:
```diff
@@ -199,6 +199,13 @@ func main() {
 	}
 	//+kubebuilder:scaffold:builder
```
You then need to remove them here. Those tell kubebuilder where to add the next controller and it might add it twice :D

QuentinBisson commented on internal/controller/dashboard_controller.go:
```diff
@@ -0,0 +1,345 @@
+	return UID, nil
+}
+
+func getDashboardCMOrg(dashboard *v1.ConfigMap) (string, error) {
```
```suggestion
func getOrgFromDashboardConfigmap(dashboard *v1.ConfigMap) (string, error) {
```

QuentinBisson commented:
Really nice work on that PR @hervenicol :)

QuentinBisson commented:
This looks really good apart from the issue around the organizations label :)

QuentinBisson commented on README.md:
```diff
@@ -7,6 +7,18 @@ This operator is in charge of handling the setup and configuration of the Giant
+
+It will look for kubernetes `ConfigMaps` and use them as dashboards if they meet these criteria:
+- a label `app.giantswarm.io/kind: "dashboard"`
+- an annotation or label `giantswarm.io/organization` set to the organization the dasboard should be loaded in.
```
```suggestion
- an annotation or label `observability.giantswarm.io/organization` set to the organization the dasboard should be loaded in.
```

Then we should be set :)

QuentinBisson commented:
Thanks for putting through with us that was some quite good go code
